File: src/main/java/dev/robothanzo/werewolf/model/Candidate.java
package dev.robothanzo.werewolf.model;

import dev.robothanzo.werewolf.database.documents.Session;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;
import org.jetbrains.annotations.Nullable;

import java.util.Collection;
import java.util.Comparator;
import java.util.LinkedList;
import java.util.List;

@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class Candidate {
    private Session.Player player;
    @Builder.Default
    private boolean expelPK = false;
    @Builder.Default
    private List<Long> electors = new LinkedList<>();
    @Builder.Default
    private boolean quit = false;

    public static Comparator<Candidate> getComparator() {
        return Comparator.comparingInt(o -> o.getPlayer().getId());
    }

    public static List<Candidate> getWinner(Collection<Candidate> candidates, @Nullable Session.Player police) {
        List<Candidate> winners = new LinkedList<>();
        float winningVotes = 0;
        for (Candidate candidate : candidates) {
            float votes = candidate.getVotes(police);
            if (votes <= 0)
                continue;
            if (votes > winningVotes) {
                winningVotes = votes;
                winners.clear();
                winners.add(candidate);
            } else if (votes == winningVotes) {
                winners.add(candidate);
            }
        }
        return winners;
    }

    public List<String> getElectorsAsMention() {
        List<String> result = new LinkedList<>();
        for (Long elector : electors) {
            result.add("<@!" + elector + ">");
        }
        return result;
    }

    public float getVotes(@Nullable Session.Player police) {
        boolean hasPolice = police != null;
        if (hasPolice)
            hasPolice = electors.contains(police.getUserId());
        return (float) ((electors.size() + (hasPolice ? 0.5 : 0)) * (quit ? 0 : 1));
    }
}

File: src/main/kotlin/dev/robothanzo/werewolf/model/Candidate.kt
package dev.robothanzo.werewolf.model

import dev.robothanzo.werewolf.database.documents.Player
import java.util.*

data class Candidate(
    val player: Player,
    var expelPK: Boolean = false,
    val electors: MutableList<Long> = LinkedList(),
    var quit: Boolean = false
) {
    fun getElectorsAsMention(): List<String> {
        return electors.map { "<@!$it>" }
    }

    fun getVotes(police: Player?): Float {
        var hasPolice = police != null
        if (hasPolice) {
            hasPolice = electors.contains(police!!.user?.idLong)
        }
        return ((electors.size + (if (hasPolice) 0.5 else 0.0)) * (if (quit) 0 else 1)).toFloat()
    }

    companion object {
        fun getComparator(): Comparator<Candidate> {
            return Comparator.comparingInt { o: Candidate -> o.player.id }
        }

        fun getWinner(candidates: Collection<Candidate>, police: Player?): List<Candidate> {
            val winners = LinkedList<Candidate>()
            var winningVotes = 0f
            for (candidate in candidates) {
                val votes = candidate.getVotes(police)
                if (votes <= 0) continue
                if (votes > winningVotes) {
                    winningVotes = votes
                    winners.clear()
                    winners.add(candidate)
                } else if (votes == winningVotes) {
                    winners.add(candidate)
                }
            }
            return winners
        }
    }
}
